src/inference/emcee.jl

```diff
@@ -0,0 +1,243 @@
+
+    ss = DynamicPPL.SampleFromPrior()
+    walkers = Vector{AMH.Transition}(undef, spl.alg.ensemble.n_walkers)
+
+    for (i, t) in enumerate(transition)
+        DynamicPPL.invlink!(spl.state.vi, spl)
+        spl.state.vi[spl] = identity.(t[ss])
+        DynamicPPL.link!(spl.state.vi, spl)
+        walkers[i] = AMH.Transition(spl.state.vi[spl], getlogp(spl.state.vi))
+    end
+    
+    return walkers
+end
+
+function AbstractMCMC.step!(
+    rng::AbstractRNG,
+    model::Model,
+    spl::Sampler{<:Emcee},
+    N::Integer,
+    transition;
+    kwargs...
+)
+    # Generate a log joint function.
+    densitymodel = AMH.DensityModel(gen_logπ(spl.state.vi, spl, model))
+
+    # Make the first transition.
+    new_transitions = AbstractMCMC.step!(rng, densitymodel, spl.alg.ensemble, 1, transition)
+    return new_transitions
+end
+
+function transform_transition(spl::Sampler{<:Emcee}, ts, w::Int, i::Int; linked=true)
+    trans = ts[i][w]
+    linked && DynamicPPL.link!(spl.state.vi, spl)
+    spl.state.vi[spl] = trans.params
+    linked && DynamicPPL.invlink!(spl.state.vi, spl)
+    setlogp!(spl.state.vi, trans.lp)
+
+    return Transition(spl)
+end
+
+function AbstractMCMC.bundle_samples(
+    rng::AbstractRNG,
+    model::AbstractModel,
+    spl::Sampler{<:Emcee},
+    N::Integer,
+    ts::Vector,
+    chain_type::Type{MCMCChains.Chains};
+    save_state = false,
+    kwargs...
+)
+    # Transform the transitions by linking them to the constrained space.
+    ts_transform = map(
+        w -> map(i -> transform_transition(spl, ts, w, i), 1:N),
+        1:spl.alg.ensemble.n_walkers
+    )
+
+    # Convert transitions to array format.
+    # Also retrieve the variable names.
+    params_vec = map(_params_to_array, ts_transform)
+
+    # Extract names and values separately.
+    nms = params_vec[1][1]
+    vals_vec = [p[2] for p in params_vec]
+
+    # Get the values of the extra parameters in each transition.
+    extra_vec = map(get_transition_extras, ts_transform)
+
+    # Get the extra parameter names & values.
+    extra_params = extra_vec[1][1]
+    extra_values_vec = [e[2] for e in extra_vec]
+
+    # Extract names & construct param array.
+    nms = [nms; extra_params]
+    parray = map(x -> hcat(x[1], x[2]), zip(vals_vec, extra_values_vec))
+    parray = cat(parray..., dims=3)
+
+    # Get the average or final log evidence, if it exists.
+    le = getlogevidence(spl)
+
+    # Set up the info tuple.
+    if save_state
+        info = (range = rng, model = model, spl = spl)
+    else
+        info = NamedTuple()
+    end
+
+    # Concretize the array before giving it to MCMCChains.
+    parray = MCMCChains.concretize(parray)
+
+    # Chain construction.
+    return MCMCChains.Chains(
+        parray,
+        string.(nms),
+        deepcopy(TURING_INTERNAL_VARS);
+        evidence=le,
+        info=info,
+        sorted=true
+    )
+end
```

Comment:
IMO it would be nice if we don't have to reimplement (most parts of) the logic for bundling samples in the end, and I guess it would be nice (and consistent with AdvancedMH and emcee) to return a Chains object with `nwalkers` numbers of chains instead of concatenating everything. The advantage would also be that the `iteration` column would not be needed.

Maybe it would be sufficient to implement `_params_to_array` for `Vector{<:Vector{<:AMH.Transition}}` (maybe one has to add a `sampler` input argument for linking and implement some other functions called in the default implementation of `bundle_samples` as well...)?

Reply:
This function does that -- it parses all the walkers out into separate chains. But I suppose the `_params_to_array` fix is a little more succinct. Although, in this case, one of the big time sucks is running `transform_transitions` to convert the walkers to the constrained space. Perhaps that should be in `sample_end!`, though.

Reply:
> Although, in this case, one of the big time sucks is running transform_transitions to convert the walkers to the constrained space. Perhaps that should be in sample_end!, though.

I'm not sure, maybe you could do that in `_params_to_array` but on the other you probably always want to get the untransformed samples back, no matter if you output a `Chains` or a vector of named tuples.

I was just surprised that you would need to implement `bundle_samples` from scratch, I would have assumed `_params_to_array` for vectors of vectors of `AMH.Transition`s (and maybe `get_logevidence`) would be sufficient since as soon as you have the transitions as a 3d array everything else should be the same.